R/SIRD_plot.R

```diff
@@ -1,121 +1,158 @@
+    dR <- (parameters["gamma"] * I) - (parameters["muD"] * R)
+    dD <- (parameters["delta"] * I)
+    dN <- dS + dI + dR
+    list(c(dS, dI, dR, dD, dN, q))
+}
+
+defaultParams <- list(
+    beta = 0.001,
+    gamma = 0.1,
+    delta = 0.05,
+    muB = 0,
+    muD = 0,
+    population = 500,
+    susceptible = 499,
+    infected = 1,
+    recovered = 0,
+    dead = 0,
+    timesteps = 50,
+    q = 0
+)
+
+# For standalone usage
+# TODO: find out if there's a better way to check if the script is being 
+# run standalone or as part of a shiny app
+solveSIRD <- function() {
+    params <- defaultParams
 
-      validate(
-        need(input$betaSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$gammaSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$deltaSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$populationSIRD > 0, "Enter a valid value for parameters"),
-        need(input$susceptibleSIRD > 0, "Enter a valid value for parameters"),
-        need(input$infectedSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$recoveredSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$deadSIRD >= 0, "Enter a valid value for parameters"),
-        need(input$timesteps > 0, "Timesteps must be greater than 0."),
-        need(input$populationSIRD == (input$susceptibleSIRD + input$infectedSIRD + input$recoveredSIRD), "Number of persons living in all epidemic compartments must sum to the Population Count (N)")
-      )
+    source <- if (!shiny::isRunning()) params else input
 
-      ode(
+    timesteps <- source$timesteps
+    beta <- source$beta
+    gamma <- source$gamma
+    delta <- source$delta
+    muB <- source$muB
+    muD <- source$muD
+    population <- source$population
+    susceptible <- source$susceptible
+    infected <- source$infected
+    recovered <- source$recovered
+    dead <- source$dead
+    q <- source$q
+
+    # Using lsoda
+    solnsSIRD <- lsoda(
         y = c(
-          S = input$susceptibleSIRD,
-          I = input$infectedSIRD,
-          R = input$recoveredSIRD,
-          D = input$deadSIRD,
-          N = input$populationSIRD,
-          q = as.integer(input$qValue)
+            S = susceptible,
+            I = infected,
+            R = recovered,
+            D = dead,
+            N = population,
+            q = q
         ),
-        times = seq(0, input$timesteps, by = 1),
-        func = sird_equations,
+        times = seq(0, timesteps, by = 1),
+        func = equationsSIRD,
         parms = c(
-          beta = input$betaSIRD,
-          gamma = input$gammaSIRD,
-          delta = input$deltaSIRD,
-          muB = input$muBirth,
-          muD = input$muDeath
+            beta = beta,
+            gamma = gamma,
+            delta = delta,
+            muB = muB,
+            muD = muD
         )
-      )
-    })
+    )
```

Comment:
Because you're providing the positional arguments in the same order as they are specified in the formal parameters of the lsoda function, you can remove the name of the argument here (which is the usual R style).

```r
  lsoda(
    c(
      S = susceptible,
      I = infected,
      R = recovered,
      D = dead,
      N = population,
      q = q
    ),
    seq(0, timesteps, by = 1),
    equationsSIRD,
    c(
      beta = beta,
      gamma = gamma,
      delta = delta,
      muB = muB,
      muD = muD
    )
  ) |>
    as.data.frame()
```

Reply:
Rather than assigning the function return value to `solnsSIRD`, please pipe it (`|>`) directly to `as.data.frame` to produce this function definition's return value.